SLA liquidity (#1685)
vegaprotocol/specs: protocol/0044-LIME-lp_mechanics.md

Important note on wording:

- liquidity provision / liquidity COMMITMENTs are the amount of stake a liquidity provider places as a bond on the market to earn rewards.
- the COMMITMENT is converted via a multiplicative network parameter `market.liquidity.stakeToCcyVolume` to a liquidity OBLIGATION, measured in price level x volume i.e. settlement currency of the market.

## Network and market parameters

### Network parameters

- `market.liquidity.bondPenaltyParameter` - used to calculate the penalty to liquidity providers when they fail to support their open position through sufficient `general+margin` balance.
Valid values: any decimal number `>= 0` with a default value of `0.1`.
- `market.liquidity.sla.nonPerformanceBondPenaltySlope` - used to calculate how much is the LP bond slashed if they fail to reach the minimum SLA. Valid values: any decimal number `>= 0` with a default value of `2.0`.
- `market.liquidity.sla.nonPerformanceBondPenaltyMax` - used to calculate how much is the LP bond slashed if they fail to reach the minimum SLA. Valid values: any decimal number `>= 0` and `<=1.0` with a default value of `0.5`.

> - `market.liquidity.maximumLiquidityFeeFactorLevel` - used in validating fee amounts that are submitted as part of the LP commitment transaction. Note that a value of `0.05 = 5%`. Valid values are: any decimal number `>0` and `<=1`. Default value `1`.

Author: ```suggestion
- `market.liquidity.maximumLiquidityFeeFactorLevel` - used in validating fee amounts that are submitted as part of the LP commitment transaction. Note that a value of `0.05 = 5%`. Valid values are: any decimal number `>=0` and `<=1`. Default value `1`.
```

- `market.liquidity.stakeToCcyVolume` - used to translate a commitment to an obligation. Any decimal number `>0` with default value `1.0`.
- `validators.epoch.length` - LP rewards from liquidity fees are paid out once per epoch according to whether they met the "SLA" (implied by `market.liquidity.committmentMinTimeFraction`) and their previous performance (for the last n epochs defined by `market.liqudity.performanceHysteresisEpochs`), see [epoch spec](./0050-EPOC-epochs.md).

> - `market.liquidity.earlyExitPenalty` (decimal between 0 and 1 inclusive), sets how much LP forfeits of their bond in case the market is below target stake and they wish to reduce their commitment. If set to `0` there is no penalty for early exit, if set to `1` their entire bond is forfeited. 

Reviewer: Ah I guess this plays into my concern above!

- `market.liquidity.probabilityOfTrading.tau.scaling` sets how the probability of trading is calculated from the risk model; this is used to [measure the relative competitiveness of LPs supplied volume](0042-LIQF-setting_fees_and_rewarding_lps.md).
- `market.liquidity.minimum.probabilityOfTrading.lpOrders` sets a lower bound on the result of the probability of trading calculation.

### Market parameters

- `market.liquidity.priceRange` (decimal) - this is a percentage price move (e.g. `0.05 = 5%`) from `mid_price` during continuous trading or indicative uncrossing price during auctions. This is set / can be modified as part of [market proposal](0028-GOVE-governance.md) / market change proposal.

- `market.liquidity.committmentMinTimeFraction` (decimal) —  minimum fraction of time LPs must spend "on the book" providing their committed liquidity. This is a decimal number in the interval $[0,1]$ i.e. both limits included. When set to $0$ the SLA mechanics are switched off for the market entirely.

> - `market.liquidity.providers.fee.calculationTimeStep` (time period e.g. `1m`) controls how often LP the quality of liquidity supplied by the LPs is evaluated and fees arising from that period are earmarked for specific parties.

Reviewer 2: ```suggestion
- `market.liquidity.providers.fee.calculationTimeStep` (time period e.g. `1m`) controls how often the quality of liquidity supplied by the LPs is evaluated and fees arising from that period are earmarked for specific parties.
```

> - `market.liquidity.providers.fee.calculationTimeStep` (time period e.g. `12h` or `7d`) controls how often LP the quality of liquidity supplied by the LPs is evaluated and fees arising from that period are earmarked for specific parties. 

Author: ```suggestion
- `market.liquidity.providers.fee.calculationTimeStep` (time period e.g. `1m`) controls how often LP the quality of liquidity supplied by the LPs is evaluated and fees arising from that period are earmarked for specific parties. 
```

  - Transfers to/from this account also occur when it is used for settlement or margin shortfall, when penalties are applied, and if the account is under-collateralised because of these uses and is subsequently topped up to the commitment amount during collateral search (see below)
- Collateral withdrawn from this account may only be transferred to either:
  - The insurance pool of the market (in event of penalties/slashing)
  - The liquidity provider's margin account or the network's settlement account/other participant's margin accounts (during a margin search and mark to market settlement) in the event that they have zero balance in their general account.
  - The liquidity provider's general account (in event of liquidity provider reducing their commitment)

### Liquidity provider proposes to amend commitment amount

The commitment transaction is also used to amend any aspect of their liquidity provision obligations.
A participant may apply to amend their commitment amount by submitting a transaction for the market with a revised commitment amount.

`proposed-commitment-variation = new-proposed-commitment-amount - old-commitment-amount`

> The amendment is actioned at the beginning of the next epoch (after the rewards / penalties for present LPs - including the party that's amending - have been evaluated). It's the amending LP's responsibility to have sufficient balance in their general account at the epoch boundary; if not the LP amendment is not actioned.

Reviewer: Thought on this, only allowing amending liquidity provisions at the end of the epoch is going to mean, for all intents and purposes and without the OI reducing stages in an auction, that dropping into a liquidity auction often means we stay there until the end of the epoch before we even have a chance to exit right? And if we miss that we wait until the next to try again?

Author: Good point, we need to fix this @barnabee.

Author: Done.

> The amendment is actioned at the beginning of the next epoch (after the rewards / penalties for present LPs - including the party that's amending - have been evaluated). It's the amending LP's responsibility to have sufficient balance in their general account at the epoch boundary; if not the LP amendment is not actioned.

Author: ```suggestion
The amendment is actioned in two steps. 
1) the amount is immediately transferred from the party's general account to a temporary "pending" bond account. This amount counts towards the stake committed to the market and so in particular can get the market out of liquidity auction. 
2) at the beginning of the next epoch (after the rewards / penalties for present LPs - including the party that's amending - have been evaluated) the amount is transferred from the "pending" bond to the true bond account. ```

Reviewer: Do they earn fees on this? I would assume they must otherwise why would someone do it, but then they could presumably increase them decrease again just before the epoch ends?

Reviewer: This also could be a pain to track if someone added multiple times in an epoch. But I can't think of much else to allow intra-epoch adding without some sort of secondary account, hmm

Reviewer: Thinking more, it might just work allowing increases whenever and decreases only at epoch end.

- Liquidity fees themselves are calculated and assigned at time of each trade, so you don't get a claim over historic fees in the epoch by submitting later
- SLA is based on your time on the book. We should ensure that if someone places their first commitment mid way through the block it counts as them having met their SLA perfectly until that point.
- Bonus payments are weighted by total fees received, so increasing later in the epoch should make that proportionally smaller. However it means your performance score is much more likely to be perfect, so I'm not quite sure how that ends up playing off against each other.

#### Increasing commitment

_Case:_ `proposed-commitment-variation >= 0`
A liquidity provider can always increase their commitment amount as long as they have sufficient collateral in the settlement asset of the market to meet the new commitment amount.

If they do not have sufficient collateral the transaction is rejected in entirety. This means that any change in fee bid is not applied and that the `old-commitment-amount` is retained.

#### Decreasing commitment

_Case:_ `proposed-commitment-variation < 0`

> We to calculate how much the LP can reduce commitment without incurring a penalty. 

Reviewer: Writing the ACs for this a further thought has come up. If we now only process these reductions in commitment at the end of an epoch there's a question of sequencing and thus penalties which might not be immediately obvious when first placing your reduction transaction (e.g. if someone else reduces their commitment such that the stake is much closer to target than it was when you first placed your reduction). 

The 'easiest' solution would just be to say we action these in the order they are placed, and LPs should just take this into account when they place their reductions, but we should probably then at least be explicit somewhere that reductions are executed in order of receipt.

Author: @TomMcL that's a good point, lets add a line saying that reductions are processed in the sequence they were placed; for each LP it's the most recent that counts.

  - Transfers to/from this account also occur when it is used for settlement or margin shortfall, when penalties are applied, and if the account is under-collateralised because of these uses and is subsequently topped up to the commitment amount during collateral search (see below)
- Collateral withdrawn from this account may only be transferred to either:
  - The insurance pool of the market (in event of penalties/slashing)
  - The liquidity provider's margin account or the network's settlement account/other participant's margin accounts (during a margin search and mark to market settlement) in the event that they have zero balance in their general account.
  - The liquidity provider's general account (in event of liquidity provider reducing their commitment)

### Liquidity provider proposes to amend commitment amount

The commitment transaction is also used to amend any aspect of their liquidity provision obligations.
A participant may apply to amend their commitment amount by submitting a transaction for the market with a revised commitment amount.

`proposed-commitment-variation = new-proposed-commitment-amount - old-commitment-amount`

The amendment is actioned in two steps.

> 1) the amount is immediately transferred from the party's general account to a temporary "pending" bond account. This amount counts towards the stake committed to the market and so in particular can get the market out of liquidity auction.

Reviewer 2: Do we really need another type of account? Could it not just go directly into the bond account with the understanding that the commitment only becomes binding at the beginning of the next epoch?

Author: Accounts are good, double entry account enforced, transfers used to move.

2) at the beginning of the next epoch (after the rewards / penalties for present LPs - including the party that's amending - have been evaluated) the amount is transferred from the "pending" bond to the true bond account.

Commitment amendments should be processed in the time order of their latest updates within the epoch. For example if an LP places an amendment at time `t1`, followed by LP2 placing an amendment at `t2` and the first LP amending again at `t1`, LP2's amendment will be executed first.

#### Increasing commitment

_Case:_ `proposed-commitment-variation >= 0`
A liquidity provider can always increase their commitment amount as long as they have sufficient collateral in the settlement asset of the market to meet the new commitment amount.

If they do not have sufficient collateral the transaction is rejected in entirety. This means that any change in fee bid is not applied and that the `old-commitment-amount` is retained.

#### Decreasing commitment

_Case:_ `proposed-commitment-variation < 0`
We to calculate how much the LP can reduce commitment without incurring a penalty.

> To do this we first evaluate the maximum amount that the market can reduce without penalty by given the current liquidity demand in the market.

Reviewer 2: ```suggestion
We to calculate how much the LP can reduce commitment without incurring a penalty.
To do this we first evaluate the maximum amount that the market can reduce without penalty given by the current liquidity demand in the market.
```

`maximum-penalty-free-reduction-amount = total_stake - target_stake`

where:

- `total_stake` is the sum of all stake of all liquidity providers bonded to this market.
- `target_stake` is a measure of the market's current stake requirements, as per the calculation in the [target stake](./0041-TSTK-target_stake.md).

If `-1 * proposed-commitment-variation <= maximum-penalty-free-reduction-amount` then we're done, the LP reduced commitment, the entire amount by which they decreased their commitment is transferred to their general account, their ELS got updated as per the [ELS calculation](0042-LIQF-setting_fees_and_rewarding_lps.md).

If `-1 * proposed-commitment-variation > maximum-penalty-free-reduction-amount` then first establish

```text
penalty-incuring-reduction-amount = -1 * proposed-commitment-variation - maximum-penalty-free-reduction-amount
```

Transfer `maximum-penalty-free-reduction-amount` to their general account.

> Now transfer `min((1-market.liquidity.earlyExitPenalty) x penalty-incuring-reduction-amount, bond account balance remaining)` to their general account and transfer `market.liquidity.earlyExitPenalty x penalty-incuring-reduction-amount` to the market insurance pool.

Reviewer 2: ```suggestion
Now transfer `min((1-market.liquidity.earlyExitPenalty) x penalty-incuring-reduction-amount, bond account balance remaining)` to their general account and transfer `market.liquidity.earlyExitPenalty x min((1-market.liquidity.earlyExitPenalty) x penalty-incuring-reduction-amount, bond account balance remaining)` to the market insurance pool.
```

The multiple is controlled by a network parameter `market.liquidity.stakeToCcyVolume`:
```
liquidity_required = stake_amount ✖️ market.liquidity.stakeToCcyVolume
```

### Meeting the committed volume of notional 

#### During continuous trading
If there is no mid price then no LP is meeting their committed volume of notional. 
If there is mid price then we calculate the volume of notional that is in the range
```
(1.0-market.liquidity.priceRange) x mid =< price levels that count <= (1+market.liquidity.priceRange)x mid.
```
If this is greater than or equal to `liquidity_required` then the LP is meeting the committed volume of notional.

#### During auctions 
We calculate the volume of notional that is in the range
```
(1.0-market.liquidity.priceRange) x min(last trade price, indicative uncrossing price) =<  price levels that count <= (1.0+market.liquidity.priceRange) x max(last trade price, indicative uncrossing price). 
```
If this is greater than or equal to `liquidity_required` then the LP is meeting the committed volume of notional.

Note: we don't evaluate whether LPs meet the SLA during opening auctions so there will always be a mark price. 

> ### Penalty for not meeting the SLA 

Reviewer: Is there anything stopping an LP near the end of the epoch who hasn’t met the SLA just withdrawing their stake (assuming over target stake) and avoiding any penalty?

Reviewer 3: Good point!

There are three options I can think of:

1. Unforced (i.e. not by bond confiscation) commitment reductions only take effect at epoch end. New commitments take effect at epoch start, too, I guess to avoid unfair penalties.

2. Your SLA is evaluated at epoch end for the time period for which you still had a commitment. This is a bit nasty because what if you have multiple different commitment levels during an epoch.

3. Yolo! Leave it and LPs can deal with it.

I'm leaning towards 1. Thoughts @TomMcL @davidsiska-vega @witgaw ?

Reviewer 2: 1 sounds nice and clean to me

When calculating fees for a trade, the size of a liquidity provider’s commitment along with when they committed and the market size are inputs that will be used to calculate how the liquidity fee is distributed between liquidity providers. See [setting fees and rewarding lps](./0042-LIQF-setting_fees_and_rewarding_lps.md) for the calculation of the split.

## Liquidity provision and penalties

### Calculating liquidity from commitment

> Committed Liquidity Providers are required to provide a multiple of their stake (supplied in the settlement currency of the market) in notional volume of orders within the range defined below.

Reviewer 2: So it's no longer per book side? That is an LP which posts only buy orders within the LP range now fully meets their commitment?

> Committed Liquidity Providers are required to provide a multiple of their stake (supplied in the settlement currency of the market) in notional volume of orders within the range defined below.

Reviewer 2: ```suggestion
Committed Liquidity Providers are required to provide a multiple of their stake (supplied in the settlement currency of the market) in notional volume of orders within the range defined below on both sides of the order book.
```

The multiple is controlled by a network parameter `market.liquidity.stakeToCcyVolume`:

```text
liquidity_required = stake_amount ✖️ market.liquidity.stakeToCcyVolume
```

### Meeting the committed volume of notional

#### During continuous trading

If there is no mid price then no LP is meeting their committed volume of notional.
If there is mid price then we calculate the volume of notional that is in the range

```text

> (1.0-market.liquidity.priceRange) x mid =< price levels that count <= (1+market.liquidity.priceRange)x mid.

Reviewer 2: ```suggestion
(1.0-market.liquidity.priceRange) x mid <= price levels <= (1+market.liquidity.priceRange)x mid.
```

```

If this is greater than or equal to `liquidity_required` then the LP is meeting the committed volume of notional.

#### During auctions

We calculate the volume of notional that is in the range

```text

> (1.0-market.liquidity.priceRange) x min(last trade price, indicative uncrossing price) =<  price levels that count <= (1.0+market.liquidity.priceRange) x max(last trade price, indicative uncrossing price).

Reviewer 2: ```suggestion
(1.0-market.liquidity.priceRange) x min(last trade price, indicative uncrossing price) <=  price levels <= (1.0+market.liquidity.priceRange) x max(last trade price, indicative uncrossing price).
```

Note:

> - As mentioned above, closeout should happen as per regular trader account (with the addition of cancelling the liquidity provision and the associated LP rewards & fees consequences). So, if after cancelling all open orders the party can afford to keep the open positions sufficiently collateralised they should be left open, otherwise the positions should get liquidated.

Reviewer 2: Is the early exit penalty applies in that case shouldn't we still charge it from the margin account?

### Bond account top up by collateral search

In the same way that a collateral search is initiated to attempt to top-up a trader's margin account if it drops below the _collateral search level_, the system must also attempt to top up the bond account if its balance drops below the size of the LP's commitment.

This should happen every time the network is performing a margin calculation and search. The system should prioritise topping up the margin account first, and then the bond account, if there are insufficient funds to do both.

If, at the end of an epoch (and before any LP rewards and penalties were applied), the LPs bond account isn't topped up to the level of their commitment then, update the ELS as per [Calculating liquidity provider equity-like share section in 0042-LIQF](./0042-LIQF-setting_fees_and_rewarding_lps.md).

## What data do we keep relating to liquidity provision?

> 1. List of all liquidity providers and their commitment sizes and their “equity-like share” for each market [see 0042-setting-fees-and-rewarding-lps](./0042-LIQF-setting_fees_and_rewarding_lps.md)

Reviewer 2: ```suggestion
1. List of all liquidity providers and their commitment sizes, their “equity-like share” and "liquidity score" for each market [see 0042-setting-fees-and-rewarding-lps](./0042-LIQF-setting_fees_and_rewarding_lps.md)
```

- If a liquidity provider with an active liquidity provision at the start of an epoch reduces their liquidity provision staked commitment during the epoch the initial committed level at the start of the epoch will remain in effect until the end of the epoch, at which point the protocol will attempt to reduce the bond to the new level. (<a name="0044-LIME-018" href="#0044-LIME-018">0044-LIME-018</a>)
- If a liquidity provider with an active liquidity provision at the start of an epoch reduces their liquidity provision staked commitment during the epoch the initial committed level at the start of the epoch will remain in effect until the end of the epoch, at which point the protocol will attempt to reduce the bond to the new level. If the reduced level has been changed several times during an epoch, only the latest value will take effect (<a name="0044-LIME-019" href="#0044-LIME-019">0044-LIME-019</a>)

> - If a liquidity provider with an active liquidity provision at the start of an epoch reduces their liquidity provision staked commitment during the epoch the initial committed level at the start of the epoch will remain in effect until the end of the epoch, at which point the protocol will attempt to reduce the bond to the new level. If the bond stake has been slashed to a level lower than the amendment, this slashed level will be retained (i.e. the protocol will not attempt to now increase the commitment) (<a name="0044-LIME-020" href="#0044-LIME-020">0044-LIME-020</a>)

Reviewer 2: Doesn't this contradict paragraph at line 253? It's not sufficiently clear to be when do we change their ELS and when (if at all) we change their commitment size.

Reviewer 2: TODO: bond account balance == commitment amount ALWAYS

- spell it out and delete paragraph about bond search stuff,
- whenever we search bond account, look into temp bond account first,

- A liquidity provider who reduces their liquidity provision such that the total stake on the market is still above the target stake after reduction will have no penalty applied and will receive their full reduction in stake back at the end of the epoch. (<a name="0044-LIME-022" href="#0044-LIME-022">0044-LIME-022</a>)
- For a market with `market.liquidity.earlyExitPenalty = 0.25` and `target stake < total stake` already, a liquidity provider who reduces their commitment by `100` will only receive `75` back into their general account with `25` transferred into the market's insurance account. (<a name="0044-LIME-023" href="#0044-LIME-023">0044-LIME-023</a>)
- For a market with `market.liquidity.earlyExitPenalty = 0.25` and `total stake = target stake + 40` already, a liquidity provider who reduces their commitment by `100` will receive a total of `85` back into their general account with `15` transferred into the market's insurance account (`40` received without penalty, then the remaining `60` receiving a `25%` penalty). (<a name="0044-LIME-024" href="#0044-LIME-024">0044-LIME-024</a>)

> - For a market with `market.liquidity.earlyExitPenalty = 0.25` and `total stake = target stake + 140` already, if one liquidity provider places a transaction to reduce their stake by `100` followed by a second liquidity provider who reduces their commitment by `100`, the first liquidity provider will receive a full `100` stake back whilst the second will receive a total of `85` back into their general account with `15` transferred into the market's insurance account (`40` received without penalty, then the remaining `60` receiving a `25%` penalty). (<a name="0044-LIME-025" href="#0044-LIME-025">0044-LIME-025</a>)

Reviewer 2: Wouldn't it make sense to only assess this at the end of epoch and apply pro rata the penalty free amounts among those wishing to leave?

I know it's an additional complication, but wondering if without it we wouldn't have LPs rushing towards the exit at certain times to avoid the penalty or even spamming the network with commitment reduction transaction (to reserve a penalty-free reduction spot for themselves) which they'll then cancel shortly before the end of an epoch if they decide to stay on.

Reviewer 2: TODO: Pro-rata (add to spec)
